Kidney Epigenome and Transcriptome-based Multi-stage Prioritization for Kidney Disease
======================================================================================

We define the genetic architecture of kidney function in 1.5 million subjects and identify 727 (136 novel) loci. We map the genotype effect on the methylome (methylation quantitative trait mQTL) in 443 human kidneys, on the transcriptome (eQTL) in 686 samples, and on single cell open chromatin in 57,282 human kidney cells.
We present a multi-stage prioritization strategy based on Bayesian multiple-trait colocalization, mendelian randomization, single cell co-accessibility and enhancer-promoter contacts. We report that methylation variation explains a larger fraction of GWAS heritability than gene expression. We assign target genes for 87% of GWAS loci, but the closest gene was only prioritized 30% of the time. We highlight the key role of proximal tubules and metabolism in kidney function regulation. We further define the causal role of _SLC47A1_ in kidney disease in mice with genetic loss of _Slc47a1_ and in human subjects carrying loss-of-function variants. 


Directory structure in this GitHub repository
----------------------------------------------
- **Kidney_mQTL** *~~~~~~~~Custom code for mQTL mapping and analysis.*
- **eGFR_GWAS** *~~~~~~~~Custom code for meta-analysis of eGFR GWAS.*
- **Kidney_eQTL** *~~~~~~~~Custom code for meta-analysis of kidney eQTL.*
- **Kidney_snATAC** *~~~~~~Custom code for snATAC data processing and anlaysis.*
- **GWAS_heritability** *~~~~Custom code for heritability estimation and analysis.*
- **Gene_prioritization** *~~~Custom code for kidney disease target gene prioritization.*


Useful links
------------------
- Susztaklab Kidney Biobank: https://susztaklab.com
- Susztak Lab: https://www.med.upenn.edu/susztaklab
